src/zalsa.rs

```diff
@@ -290,15 +293,39 @@ impl Zalsa {
 
     pub(crate) fn ingredient_index_for_memo(
         &self,
+        struct_ingredient_index: IngredientIndex,
         memo_ingredient_index: MemoIngredientIndex,
     ) -> IngredientIndex {
-        self.memo_ingredients.lock()[memo_ingredient_index.as_usize()]
+        self.memo_ingredient_indices.read()[struct_ingredient_index.as_usize()]
+            [memo_ingredient_index.as_usize()]
     }
 }
 
-impl JarAux for Zalsa {
-    fn next_memo_ingredient_index(&self, ingredient_index: IngredientIndex) -> MemoIngredientIndex {
-        let mut memo_ingredients = self.memo_ingredients.lock();
+struct JarAuxImpl<'a>(&'a Zalsa, &'a FxHashMap<TypeId, IngredientIndex>);
```

Comment:
I changed salsa struct ingredient lookup as I said [here](https://github.com/salsa-rs/salsa/pull/614#discussion_r1841483937) but I had to change this to avoid deadlocks 😢